Sources/SourceKitLSP/Swift/MacroExpansion.swift

```diff
@@ -0,0 +1,94 @@
+
+/// Detailed information about the result of a macro expansion operation.
+///
+/// Wraps the information returned by sourcekitd's `semantic_refactoring` request, such as the necessary macro expansion edits.
+struct MacroExpansion: Refactoring {
+
+  /// The title of the refactoring action.
+  var title: String
+
+  /// The URI of the file where the macro is used
+  var uri: DocumentURI
+
+  /// The resulting array of `MacroExpansionEdit` of a semantic refactoring request
+  var edits: [MacroExpansionEdit]
+
+  init(title: String, uri: DocumentURI, refactoringEdits: [RefactoringEdit]) {
+    self.title = title
+    self.uri = uri
+    self.edits = refactoringEdits.map { refactoringEdit in
+      MacroExpansionEdit(
+        range: refactoringEdit.startPosition..<refactoringEdit.endPosition,
+        newText: refactoringEdit.newText,
+        bufferName: refactoringEdit.bufferName ?? ""  // TODO: handle bufferName if nil
+      )
+    }
+  }
+
+}
+
+extension SwiftLanguageService {
+  /// Handles the `ExpandMacroCommand`.
+  ///
+  /// Makes request to sourcekitd and wraps the result into a `MacroExpansion` and then makes `ShowDocumentRequest` to the client side for each expansion to be displayed.
+  ///
+  /// - Parameters:
+  ///   - refactorCommand: The expand macro `SwiftCommand` that triggered this request.
+  ///
+  /// - Returns:
+  ///   - an `[MacroExpansionEdit]` with the necessary edits and buffer name as a `LSPAny`
+  func expandMacro(
+    _ refactorCommand: any SwiftCommand
+  ) async throws -> LSPAny {
+    guard let sourceKitLSPServer else {
+      // `SourceKitLSPServer` has been destructed. We are tearing down the language
+      // server. Nothing left to do.
+      throw ResponseError.unknown("Connection to the editor closed")
+    }
+
+    guard let refactorCommand = refactorCommand as? ExpandMacroCommand else {
+      throw ResponseError.unknown("refactorCommand is not a ExpandMacroCommand")
+    }
+
+    let expansion = try await self.refactoring(refactorCommand, MacroExpansion.self)
+
+    for macroEdit in expansion.edits {
+      let macroExpansionFilePath = self.generatedMacroExpansionsPath.appendingPathComponent(
+        macroEdit.bufferName
+      )
+      let macroExpansionDocURI = DocumentURI(macroExpansionFilePath)
+      if let _ = try? self.documentManager.latestSnapshot(macroExpansionDocURI) {
+        continue
+      }
+
+      do {
+        try macroEdit.newText.write(to: macroExpansionFilePath, atomically: true, encoding: String.Encoding.utf8)
+      } catch {
+        throw ResponseError.unknown("Unable to write macro expansion to file path: \"\(macroExpansionFilePath.path)\"")
+      }
+
+      let req = ShowDocumentRequest(uri: macroExpansionDocURI, selection: macroEdit.range)
+      let response = try await sourceKitLSPServer.sendRequestToClient(req)
+      if !response.success {
+        logger.error("client refused to show document for \(expansion.title, privacy: .public)!")
+      }
```

Comment:
I would prefer to send this request to the client asynchronously (ie. wrap it in a `Task`) because otherwise it’s an easy way to deadlock SourceKit-LSP in the future. I don’t think it can happen right now but depending on how `MessageHandlingDependencyTracker` gets changed in the future it’s not completely unreasonable to assume that as part of the `ShowDocumentRequest` to the client, the client will send a request to SourceKit-LSP that cannot be executed until the `ExecuteCommandRequest` finishes executing. But that can’t finish executing until the `ShowDocumentRequest` has finished, so deadlock.

Reply:
Yeah that makes sense